README.md
HASH Library
============

## About
Main purpose of the Library is **encapsulate** hashing mechanisms and **give** to them a convenient and simple interface of access:

```php
// give default strategy (task: HASH::COMMON, strategy: HASH::MD5)
$crypt = HASH::getInstance();

// give default strategy for password hashing (strategy: HASH::SHA1_MD5)
$crypt = HASH::getInstance(HASH::PASSWORD);

// give selected strategy for password hashing
$crypt = HASH::getInstance(HASH::PASSWORD, array(
	'strategy' => HASH::MD5_SALT_SHA1,		// some strategies require salt, by default $salt = substr(md5(__CLASS__), 0, 22)
	'salt' => $this->config->item('salt'),	// salt, for example, can be stored globally in the site settings
));
```


## Information
* [Official page](http://www.octolab.org/libraries/hash)
* You can read a useful [article](http://www.yiiframework.com/wiki/425/use-crypt-for-password-storage/) about using crypt() function for password storage (in particular about advantages of the [Blowfish algorithm](http://en.wikipedia.org/wiki/Blowfish_%28cipher%29) on Yii framework wiki pages



## Requirements
* PHP grater or equals than 5.3.0


## Preparation
Classes are connected via a **relative path**. And therefore you must add parent folder of the Library into include path:

```php
set_include_path(get_include_path() . PATH_SEPARATOR . 'path/to/folder/which/contains/HASH');
```


## Work with library
Hashing and comparing:

```php
$crypt = HASH::getInstance(HASH::PASSWORD);
$hash = $crypt->make($string);

if ($crypt->compare($input, $stored)) {
	echo 'Match';
} else {
	echo 'Do not match';
}
```


## Adapters
### Yii
Yii_Hash designed for handy integration the Library with [Yii framework](http://www.yiiframework.com/).

Setup configuration of the Library:
```php
Yii::setPathOfAlias('HASH', 'path/to/folder/which/contains/HASH'); // set namespace

return array(
	...,
	'components' => array(
		...,
		'hash' => array(
			'class' => 'HASH\adapters\YII_Hash',
			'strategies' => array(
				'pass' => array(
					'strategy' => HASH\HASH::BLOWFISH_RANDOM_SALT,
					'cost' => 12,
				),
			),
		),
	),
);
```

After that you can use it as usual component Yii:
```php
$hash = Yii::app()->hash->pass->make($input);

if (Yii::app()->hash->pass->compare($input, $stored)) {
	echo 'Match';
} else {
	echo 'Do not match';
}
```

### CodeIgniter
CI_Hash designed for handy integration the Library with [CodeIgniter framework](http://ellislab.com/codeigniter).

You can define configuration of the Library in hash.php file in config folder of CodeIgniter:

```php
$config = array(
	'pass' => array(
		'strategy' => HASH::SHA1_MD5_SALT,
		'salt' => 'q3XBgoiRCXfuTertfplXv1ICT',
	),
	'email' => array(
		'strategy' => HASH::MD5_SALT,
		'salt' => 'GswJNrMQAA_Q',
	),
);
```

Or you can put config into loader:

```php
$config = array(
	'pass' => array(
		'strategy' => HASH::SHA1_MD5_SALT,
		'salt' => 'q3XBgoiRCXfuTertfplXv1ICT',
	),
	'email' => array(
		'strategy' => HASH::MD5_SALT,
		'salt' => 'GswJNrMQAA_Q',
	),
);
$this->load->library('hash', $config);
```

After that you can use it as usual library for CodeIgniter:

```php
$hash = $this->hash->pass->make($input);

if ($this->hash->pass->compare($input, $stored)) {
	echo 'Match';
} else {
	echo 'Do not match';
}
```

### Silex
SLX_Hash designed for handy integration the Library with [Silex framework](http://silex.sensiolabs.org/).

Access to the Library implemented via ServiceProvider:

```php
use \HASH\adapters\SLX_Hash as Hash;
$app->register(new Hash(), array(
	'hash.task' => HASH::SHA1_MD5_SALT,
	'hash.config' => array(
		'strategy' => HASH::SHA1_MD5_SALT,
		'salt' => 'q3XBgoiRCXfuTertfplXv1ICT',
	),
));
```

After that you can use it as usual service provider for Silex:

```php
$pass = 'hash_of_super_secret_pass';

$app->post('login', function (Application $app, Request $request) use ($pass) {
	$crypt = $app['hash'];
	if ($crypt->compare($request->get('pass'), $pass)) {
		echo 'Match';
	} else {
		echo 'Do not match';
	}
})
```


## Unit tests
All tests located at `tests` folder. You can run it as follows:

```
% phpunit StrategiesUnitTest
% phpunit ComponentUnitTest
```


## License

Copyright (c) HASH Library from OctoLab

Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files (the "Software"), to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.


## [Changelog](CHANGELOG.md)

CHANGELOG.md
# 1.2 (April 8, 2013)
* Add Yii support
	* Yii_Hash adapter
	* Yii autoloader support
* Code refactoring

# 1.1 (And this)
* Add Blowfish algorithm
	* SimpleBlowfish (base for strategies family)
	* Blowfish
	* BlowfishSalt
	* BlowfishRandomSalt


# 1.0 (I forgot when it was)
* Define general strategies

HASH.php
<?php
/**
 * @author Samigullin Kamil <feedback@kamilsk.com>
 * @link http://www.kamilsk.com/
 */
namespace HASH;
use HASH\interfaces\iHash;
use \Exception;
/**
 * @package HASH
 * @since 1.0
 */
abstract class HASH
{
	const VERSION = 1.2;
	/**
	 * Strategies.
	 */
	const MD5 = 1;
	const MD5_SALT = 2;
	const SALT_MD5 = 3;
	const SHA1 = 4;
	const SHA1_SALT = 5;
	const SALT_SHA1 = 6;
	const MD5_SHA1 = 7;
	const SHA1_MD5 = 8;
	const MD5_SHA1_SALT = 9;
	const MD5_SALT_SHA1 = 10;
	const SALT_MD5_SHA1 = 11;
	const SHA1_MD5_SALT = 12;
	const SHA1_SALT_MD5 = 13;
	const SALT_SHA1_MD5 = 14;
	const BLOWFISH = 15;
	const BLOWFISH_SALT = 16;
	const BLOWFISH_RANDOM_SALT = 17;
	/**
	 * Tasks.
	 */
	const COMMON = 0;
	const PASSWORD = 1;
	const EMAIL = 2;
	/**
	 * @var bool
	 */
	public static $included = true;
	/**
	 * Map.
	 */
	protected static $_strategiesMap = array(
		self::MD5 => 'Md5',
		self::MD5_SALT => 'Md5Salt',
		self::SALT_MD5 => 'SaltMd5',
		self::SHA1 => 'Sha1',
		self::SHA1_SALT => 'Sha1Salt',
		self::SALT_SHA1 => 'SaltSha1',
		self::MD5_SHA1 => 'Md5Sha1',
		self::SHA1_MD5 => 'Sha1Md5',
		self::MD5_SHA1_SALT => 'Md5Sha1Salt',
		self::MD5_SALT_SHA1 => 'Md5SaltSha1',
		self::SALT_MD5_SHA1 => 'SaltMd5Sha1',
		self::SHA1_MD5_SALT => 'Sha1Md5Salt',
		self::SHA1_SALT_MD5 => 'Sha1SaltMd5',
		self::SALT_SHA1_MD5 => 'SaltSha1Md5',
		self::BLOWFISH => 'Blowfish',
		self::BLOWFISH_SALT => 'BlowfishSalt',
		self::BLOWFISH_RANDOM_SALT => 'BlowfishRandomSalt',
	);
	/**
	 * Tasks map.
	 */
	protected static $_tasksMap = array(
		self::COMMON => self::MD5,
		self::PASSWORD => self::SHA1_MD5,
		self::EMAIL => self::SHA1,
	);
	/**
	 * @var array
	 */
	protected static $_instances = array();

	/**
	 * Registry.
	 *
	 * Configuration:
	 * <code>
	 * array(
	 * 	['strategy' => HASH::MD5..HASH::SALT_SHA1_MD5,]
	 * 	['salt' => '%string',]
	 * 	['cost' => '%string',]
	 * );
	 * </code>
	 * @param int|string $task
	 * @param array $config
	 * @return iHash
	 * @throws Exception
	 */
	public static function getInstance($task = self::COMMON, array $config = array())
	{
		if ( ! is_int($task) && ( ! is_string($task) or empty($task))) {
			throw new Exception('_invalid_args');
		}
		if ( ! isset(self::$_instances[$task])) {
			$config = array_merge(
				array(
					'strategy' => isset(self::$_tasksMap[$task]) ? self::$_tasksMap[$task] : self::$_tasksMap[self::COMMON],
					'salt' => substr(md5(__CLASS__), 0, 22),
					'cost' => 11,
				),
				$config
			);
			if ( ! isset(self::$_strategiesMap[$config['strategy']])) {
				throw new Exception('_unknown_strategy');
			}
			$file = self::$_strategiesMap[$config['strategy']];
			$strategy = "HASH\\strategies\\{$file}";
			if (self::$included) {
				class_exists($strategy, false) or require "HASH/strategies/{$file}.php";
			}
			self::$_instances[$task] = new $strategy($config);
		} elseif ( ! empty($config)) {
			throw new Exception('_already_exist_and_unmodified');
		}
		return self::$_instances[$task];
	}
}